### PR TITLE
Impl drag-drop action for Windows

### PR DESCRIPTION
**Pull request by JunkuiZhang**
### Description

This is a part of #8809 

https://github.com/zed-industries/zed/assets/14981363/2b085b9d-8b83-4ac7-8b84-07c679760eba

Release Notes:

- N/A

### CODE REVIEW

File: crates/gpui/src/platform/windows/window.rs [hunk 1]
```diff
@@ -579,6 +602,7 @@ struct Callbacks {
 
 pub(crate) struct WindowsWindow {
     inner: Rc<WindowsWindowInner>,
+    dragdrop_handler: IDropTarget,
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/window.rs:605]: `dragdrop` -> `drag_drop`

File: crates/gpui/src/platform/windows/window.rs [hunk 2]
```diff
@@ -562,6 +573,18 @@ impl WindowsWindowInner {
         }
         LRESULT(1)
     }
+
+    fn handle_dragdrop(&self, input: PlatformInput) {
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/window.rs:577]: `dragdrop` -> `drag_drop`

File: crates/gpui/src/platform/windows/window.rs [hunk 3]
```diff
@@ -562,6 +573,18 @@ impl WindowsWindowInner {
...
+        let mut callbacks = self.callbacks.borrow_mut();
+        let Some(ref mut func) = callbacks.input else {
+            return;
+        };
+        func(input);
+        // TODO: remove this if we continuously draw the view
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/window.rs:583]: We should remove this now that #8921 is in.

File: crates/gpui/src/platform/windows/window.rs [hunk 4]
```diff
@@ -643,8 +667,19 @@ impl WindowsWindow {
                 lpparam,
             )
         };
+        let dragdrop_handler = {
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/window.rs:670]: `dragdrop` -> `drag_drop`

File: crates/gpui/src/platform/windows/window.rs [hunk 5]
```diff
@@ -643,8 +667,19 @@ impl WindowsWindow {
...
+            let inner = context.inner.as_ref().unwrap();
+            let handler = WindowsDragDropHandler(Rc::clone(inner));
+            let dragdrop_handler: IDropTarget = handler.into();
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/window.rs:673]: `dragdrop` -> `drag_drop`

File: crates/gpui/src/platform/windows/window.rs [hunk 6]
```diff
@@ -643,8 +667,19 @@ impl WindowsWindow {
...
+            unsafe {
+                RegisterDragDrop(inner.hwnd, &dragdrop_handler)
+                    .expect("unable to register drag-drop event")
+            };
+            dragdrop_handler
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/window.rs:678]: `dragdrop` -> `drag_drop`

File: crates/gpui/src/platform/windows/window.rs [hunk 7]
```diff
@@ -926,3 +1084,6 @@ unsafe fn set_window_long(hwnd: HWND, nindex: WINDOW_LONG_PTR_INDEX, dwnewlong:
         SetWindowLongW(hwnd, nindex, dwnewlong as i32) as isize
     }
 }
+
+const DRAGDROP_GET_COUNT: u32 = 0xFFFFFFFF;
+const FILENAME_MAXLENGTH: usize = 256;
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/window.rs:1089]: Shouldn't this be `windows::Win32::Foundation::MAX_PATH`?

File: crates/gpui/src/platform/windows/window.rs [hunk 8]
```diff
@@ -926,3 +1084,6 @@ unsafe fn set_window_long(hwnd: HWND, nindex: WINDOW_LONG_PTR_INDEX, dwnewlong:
         SetWindowLongW(hwnd, nindex, dwnewlong as i32) as isize
     }
 }
+
+const DRAGDROP_GET_COUNT: u32 = 0xFFFFFFFF;
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/window.rs:1088]: Where does this constant come from?

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

**Comment by mikayla-maki** [comment]
Looks great! If you could resolve some naming and comment issues + get the PR to passing CI and mergable, we can get this in :)

File: crates/gpui/src/platform/windows/window.rs (hunk 6 above)
Comment (JunkuiZhang) [reply crates/gpui/src/platform/windows/window.rs]: Done.

File: crates/gpui/src/platform/windows/window.rs (hunk 8 above)
Comment (JunkuiZhang) [reply crates/gpui/src/platform/windows/window.rs]: Sorry for the delay, it was 2 am when I submitted this PR.

By [Microsoft](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-dragqueryfilew):
> If the value of this parameter is 0xFFFFFFFF, DragQueryFile returns a count of the files dropped.

File: crates/gpui/src/platform/windows/window.rs (hunk 1 above)
Comment (JunkuiZhang) [reply crates/gpui/src/platform/windows/window.rs]: Done.

File: crates/gpui/src/platform/windows/window.rs (hunk 7 above)
Comment (JunkuiZhang) [reply crates/gpui/src/platform/windows/window.rs]: That's right.

**Comment by a-usr** [comment]
You could suppress this warning as documented [here](https://crates.io/crates/typos-cli#false-positives)

**Review by mikayla-maki — APPROVED** [review]

**Pull request merged**